irohad/model/converters/impl/json_query_factory.cpp

```diff
@@ -15,241 +15,106 @@
-          hexstringToArray(sig["signature"].GetString(), signature.signature);
-          query->signature = signature;
-          return query;
-        }
-        log_->error("No query type found");
-        return nonstd::nullopt;
+        return stringToJson(query_json)
+            | [this](auto &json) { return this->deserialize(json); };
       }
 
-      std::shared_ptr<iroha::model::Query>
-      JsonQueryFactory::deserializeGetAccount(
-          rapidjson::GenericValue<rapidjson::UTF8<char>>::Object &obj_query) {
-        if (not obj_query.HasMember("account_id")) {
-          log_->error("No account id in json");
-          return nullptr;
-        }
-        if (not obj_query["account_id"].IsString()) {
-          log_->error("Type mismatch account_id in json");
-          return nullptr;
-        }
-        auto timestamp = obj_query["created_ts"].GetUint64();
-        auto creator = obj_query["creator_account_id"].GetString();
-        auto counter = obj_query["query_counter"].GetUint64();
-        auto account_id = obj_query["account_id"].GetString();
-        return query_generator_.generateGetAccount(timestamp, creator, counter,
-                                                   account_id);
+      optional_ptr<Query> JsonQueryFactory::deserialize(
+          const rapidjson::Document &document) {
+        auto des = makeFieldDeserializer(document);
+        return des.String("query_type")
+            | makeOptionalGet(deserializers_)
+            | makeMethodInvoke(*this, document)
+            | des.Uint64(&Query::created_ts, "created_ts")
+            | des.String(&Query::creator_account_id, "creator_account_id")
+            | des.Uint64(&Query::query_counter, "query_counter")
+            | des.Object(&Query::signature, "signature")
+            | [this, &document](auto query) {
+                query->query_hash = hash_provider_.get_hash(query);
+                return nonstd::make_optional(query);
+              };
       }
 
-      std::shared_ptr<iroha::model::Query>
-      JsonQueryFactory::deserializeGetSignatories(
-          rapidjson::GenericValue<rapidjson::UTF8<char>>::Object &obj_query) {
-        if (not obj_query.HasMember("account_id")) {
-          log_->error("No account id in json");
-          return nullptr;
-        }
-        if (not obj_query["account_id"].IsString()) {
-          log_->error("Type mismatch account_id in json");
-          return nullptr;
-        }
-
-        auto timestamp = obj_query["created_ts"].GetUint64();
-        auto creator = obj_query["creator_account_id"].GetString();
-        auto counter = obj_query["query_counter"].GetUint64();
-        auto account_id = obj_query["account_id"].GetString();
-        return query_generator_.generateGetSignatories(timestamp, creator,
-                                                       counter, account_id);
+      optional_ptr<Query> JsonQueryFactory::deserializeGetAccount(
```

Comment:
What do you think about template function `deserializeQuery` that is defined for the following structures? Then such painful function names will dissapear

Reply:
Sure, we can fix this in another pull request for all factories.